qualtran/surface_code/azure_cost_model.py

```diff
@@ -0,0 +1,78 @@
+#  Copyright 2023 Google LLC
+#
+#  Licensed under the Apache License, Version 2.0 (the "License");
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#      https://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+
+import math
+
+import numpy as np
+
+from qualtran.surface_code.algorithm_specs import AlgorithmSpecs
+from qualtran.surface_code.magic_state_factory import MagicStateCount
+from qualtran.surface_code.physical_parameters import PhysicalParameters
+from qualtran.surface_code.quantum_error_correction_scheme import QuantumErrorCorrectionScheme
+from qualtran.surface_code.rotation_cost_model import RotationCostModel
+from qualtran.surface_code.t_factory import TFactory
+
+
+def logical_qubits(algorithm_specs: AlgorithmSpecs) -> int:
+    """Equation D1 in https://arxiv.org/abs/2211.07629"""
```

Comment:
can you document why this is the formula in the docstring and then provide the equation reference as a reference? Ideally this file could be a pretty self-contained description of a cost model